File: python/7 kyu/301.py
# https://www.codewars.com/kata/5a58d46cfd56cb4e8600009d/train/python

# Passed

def halving_sum(n): 
    total_sum = 0
    divisor = 1
    while (term := n // divisor) > 0:
        total_sum += term
        divisor *= 2
    
    return total_sum

output = halving_sum(25)
print(output)

File: nodejs/7 kyu/301.js
// https://www.codewars.com/kata/56b0f6243196b9d42d000034/train/javascript

// Passed

function sumFactorial(arr) {
  let numToFactorialMap = new Object();
  let calcFactorial = (num) => {
    if (num === 1) {
      return 1;
    }

    let factorial =
      num in numToFactorialMap
        ? numToFactorialMap[num]
        : num * calcFactorial(num - 1);
    numToFactorialMap[num] = factorial;
    return factorial;
  };

  let totalSum = 0;
  for (const num of arr) {
    let numFactorial = calcFactorial(num);
    totalSum += numFactorial;
  }

  return totalSum;
}

const output = sumFactorial([4, 6]);
console.log(output);